File: codebase/backend/src/main/java/it/unisa/scanapp/domain/Allergen.java
package it.unisa.scanapp.domain;
import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import javax.validation.constraints.*;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

/**
 * A Allergen.
 */
@Entity
@Table(name = "allergen")
public class Allergen implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "sequenceGenerator")
    @SequenceGenerator(name = "sequenceGenerator")
    private Long id;

    @NotNull
    @Column(name = "name", nullable = false)
    private String name;

    @NotNull
    @Column(name = "description", nullable = false)
    private String description;

    @Column(name = "image_url")
    private String imageUrl;

    @ManyToMany
    @JoinTable(name = "allergen_user",
               joinColumns = @JoinColumn(name = "allergen_id", referencedColumnName = "id"),
               inverseJoinColumns = @JoinColumn(name = "user_id", referencedColumnName = "id"))
    @JsonIgnore
    private Set<User> users = new HashSet<>();

    @ManyToMany(mappedBy = "allergens")
    @JsonIgnore
    private Set<Product> products = new HashSet<>();

    // jhipster-needle-entity-add-field - JHipster will add fields here, do not remove
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public Allergen name(String name) {
        this.name = name;
        return this;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public Allergen description(String description) {
        this.description = description;
        return this;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public Allergen imageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
        return this;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public Set<User> getUsers() {
        return users;
    }

    public Allergen users(Set<User> users) {
        this.users = users;
        return this;
    }

    public Allergen addUser(User user) {
        this.users.add(user);
        return this;
    }

    public Allergen removeUser(User user) {
        this.users.remove(user);
        return this;
    }

    public void setUsers(Set<User> users) {
        this.users = users;
    }

    public Set<Product> getProducts() {
        return products;
    }

    public Allergen products(Set<Product> products) {
        this.products = products;
        return this;
    }

    public Allergen addProduct(Product product) {
        this.products.add(product);
        product.getAllergens().add(this);
        return this;
    }

    public Allergen removeProduct(Product product) {
        this.products.remove(product);
        product.getAllergens().remove(this);
        return this;
    }

    public void setProducts(Set<Product> products) {
        this.products = products;
    }
    // jhipster-needle-entity-add-getters-setters - JHipster will add getters and setters here, do not remove

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Allergen)) {
            return false;
        }
        return id != null && id.equals(((Allergen) o).id);
    }

    @Override
    public int hashCode() {
        return 31;
    }

    @Override
    public String toString() {
        return "Allergen{" +
            "id=" + getId() +
            ", name='" + getName() + "'" +
            ", description='" + getDescription() + "'" +
            ", imageUrl='" + getImageUrl() + "'" +
            "}";
    }
}

File: codebase/frontend/src/app/shared/model/allergen.ts
import { IUser, User } from './user';

export interface IAllergen {
    id: number;
    name: string;
    description: string;
    imageUrl: string;
    users: IUser[];
}

export class Allergen {
    private id: number;
    private name: string;
    private description: string;
    private imageUrl: string;
    private users: Array<User>;

    constructor(allergen?: IAllergen) {
        this.users = new Array<User>();
        if (allergen) {
            this.id = allergen.id;
            this.name = allergen.name;
            this.description = allergen.description;
            this.imageUrl = allergen.imageUrl;
            if (allergen.users != null) {
                allergen.users.forEach(element => {
                    this.users.push(new User(element));
                });
            }
        }
    }

    public getId(): number {
        return this.id;
    }

    public setId(id: number): void {
        this.id = id;
    }

    public getName(): string {
        return this.name;
    }

    public setName(name: string): void {
        this.name = name;
    }

    public getDescription(): string {
        return this.description;
    }

    public setDescription(description: string): void {
        this.description = description;
    }

    public getImageUrl(): string {
        return this.imageUrl;
    }

    public setImageUrl(imageUrl: string): void {
        this.imageUrl = imageUrl;
    }
}
